Testing Universal Links (ios) (#6)
jeel38/owasp-mastg: demos/ios/MASVS-PLATFORM/MASTG-DEMO-0x70-5/MASTG-DEMO-0x70-5.md

---
platform: ios
title: Extracting the openURL Selector from the Binary
id: MASTG-DEMO-0070-5
code: [swift]
test: MASTG-TEST-0070-5
tools: [MASTG-TOOL-0129]
---

## Sample

> The source code snippet below passes an unvalidated URL to `UIApplication.shared.open`. During a black-box assessment, testers verify this behavior by extracting the imported `openURL:` Objective-C selectors from the compiled binary.

Reviewer: ```suggestion
The code snippet below demonstrates the extraction of the unvalidated URL to the `UIApplication.shared.open`  method from the binary.
```

{{ ../MASTG-DEMO-0x70-1/MastgTest.swift }}

## Steps

1. Unzip the app package and locate the main app binary (@MASTG-TECH-0058), which in this case is `./Payload/MASTestApp.app/MASTestApp`.
2. Run `run.sh` to parse the binary strings and method signatures using `rabin2` (@MASTG-TOOL-0129).

> 3. Check the output for the `openURL:options:completionHandler:` selector.

Reviewer: ```suggestion
2. Run `run.sh` to parse the binary strings and method signatures using @MASTG-TOOL-0129.
```

{{ run.sh }}

## Observation

> The output confirms that the compiled application imports and utilizes the iOS APIs responsible for opening external links.

Reviewer: Observation is not according to the guidelines. It should begin with `The output contains ...`

reference: https://github.com/OWASP/mastg/pull/3038/changes#diff-f60738d29b77815822088ff0496bb766ef3035d39a4581631e4ecd225b4966e7:~:text=MASTG%2DDEMO%2D0066-,MASTG%2DDEMO%2D0066.md,-output.txt

{{ output.txt }}

## Evaluation

The test confirms the application calls the `openURL:options:completionHandler:` selector.

> Finding this selector acts as a trigger for further investigation. The tester must dynamically trace the data flow to ensure that untrusted inbound Universal Links are not blindly passed directly into this outgoing `openURL` function, which would result in URI Scheme Hijacking.

Reviewer: Evaluation not according to the guidelines and refer the [MASTG-0x70-1](https://github.com/jeel38/owasp-mastg/pull/6/changes#diff-6358f416c8c6434c1fbda0d1eeea9b41257483e57915882b6d66478c1e2be98a) demo evaluation of the lines.  It should state the test Pass/Fails

Reference: https://github.com/OWASP/mastg/pull/3038/changes#diff-f60738d29b77815822088ff0496bb766ef3035d39a4581631e4ecd225b4966e7:~:text=MASTG%2DDEMO%2D0066-,MASTG%2DDEMO%2D0066.md,-output.txt
